sentry-actix/src/lib.rs

```diff
@@ -68,221 +71,267 @@
-        SentryMiddleware {
+    pub fn new() -> Self {
+        Sentry(Rc::new(Inner {
             hub: None,
             emit_header: false,
             capture_server_errors: true,
-        }
+        }))
     }
 
     /// Creates a new middleware builder.
-    pub fn builder() -> SentryMiddlewareBuilder {
-        SentryMiddleware::new().into_builder()
+    pub fn builder() -> SentryBuilder {
+        Sentry::new().into_builder()
     }
 
     /// Converts the middleware into a builder.
-    pub fn into_builder(self) -> SentryMiddlewareBuilder {
-        SentryMiddlewareBuilder { middleware: self }
+    pub fn into_builder(self) -> SentryBuilder {
+        SentryBuilder { middleware: self }
     }
+}
 
-    fn new_hub(&self) -> Arc<Hub> {
-        Arc::new(Hub::new_from_top(Hub::main()))
+impl Default for Sentry {
+    fn default() -> Self {
+        Sentry::new()
     }
 }
 
-impl Default for SentryMiddleware {
-    fn default() -> Self {
-        SentryMiddleware::new()
+impl<S, B> Transform<S> for Sentry
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Transform = SentryMiddleware<S>;
+    type InitError = ();
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ok(SentryMiddleware {
+            service,
+            inner: self.0.clone(),
+        })
     }
 }
 
-fn extract_request<S: 'static>(
-    req: &HttpRequest<S>,
+/// The middleware for individual services.
+pub struct SentryMiddleware<S> {
+    service: S,
+    inner: Rc<Inner>,
+}
+
+impl<S, B> Service for SentryMiddleware<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: ServiceRequest) -> Self::Future {
+        let inner = self.inner.clone();
+        let hub = Arc::new(Hub::new_from_top(Hub::main()));
+        let client = hub.client();
+        let with_pii = client
+            .as_ref()
+            .map_or(false, |x| x.options().send_default_pii);
+        let guard = hub.push_scope();
+
+        let (tx, sentry_req) = sentry_request_from_http(&req, with_pii);
+        hub.configure_scope(|scope| {
+            scope.add_event_processor(Box::new(move |event| {
+                process_event(event, tx.clone(), &sentry_req)
+            }))
+        });
+
+        req.extensions_mut().insert(HubWrapper { hub: hub.clone() });
```

Comment:
Since sentry has futures support now anyway, it might be a good idea to use that, and bind the Hub to the future.
If I understand correctly, the new actix is completely based on Futures, right? That would mean that people can use `Hub::current/with` etc and get the correct hub, instead of having an actix specific method to get the Hub.

Reply:
Since `ActixWebHubExt` was removed, there isn't a need to store the Hub in the extensions anymore. However, to make it possible for users to access the current Hub through `Hub::current`, is there a function to 'bind the Hub to the future'?

Reply:
yes, thats what I meant: https://github.com/getsentry/sentry-rust/blob/c8ef5fecea34215776863b0333a734c247a6f23b/sentry-core/src/futures.rs#L55

Reply:
Just to make sure this is the change you're suggesting:

```rust
let fut = self.service.call(req).bind_hub(hub.clone());
```